### PR TITLE
Use the same code for calculating forward and backward probabilities in CTC

### PR DESCRIPTION
**Pull request by unnonouno**
Calculation of forward and backward probabilities are very similar. I fixed matrix order and make them the same code. And after that, I combine all kernels and reduced the number of kernels. When batch size is small this code is 3x faster.
This fix also includes refactoring. I can split the PR.

### CODE REVIEW

File: chainer/functions/loss/ctc.py [hunk 1]
```diff
@@ -147,17 +193,21 @@ def _computes_transition(self, prev_prob, path, path_length):
             prob = _logsumexp(mat, xp, axis=0)
             outside = xp.arange(max_path_length) >= path_length[:, None]
             prob[outside] = self.zero_padding
+            cum_prob += prob
+            batch_index = xp.arange(0, n_batch, dtype='i')
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: How about removing `0`?

File: chainer/functions/loss/ctc.py [hunk 2]
```diff
@@ -183,48 +239,29 @@ def calc_trans(self, yseq, input_length,
...
+        forward_prob[:, 0] = 0
         backward_prob = forward_prob
-        offset = xp.arange(
-            0, n_batch * n_unit, n_unit, dtype=path.dtype)[:, None]
 
-        # prob[i] := forward[i] + backward[-i-1]
-        index = offset + path
-        prob = xp.empty(
-            (max_input_length, n_batch, max_path_length), dtype='f')
+        batch_index = xp.arange(0, n_batch, dtype='i')
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]: How about removing `0`?

**Comment by unnonouno** [comment]
updated

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**